File: src/main/java/recipeapplication/controller/WebController.java
package recipeapplication.controller;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestParam;
import recipeapplication.exception.InvalidPasswordTokenException;
import recipeapplication.exception.RecipeDoesNotExistException;
import recipeapplication.model.*;
import recipeapplication.service.AuthService;
import recipeapplication.service.InventoryService;
import recipeapplication.service.RecipeService;
import recipeapplication.service.ShoppingListService;
import recipeapplication.service.UserService;

import java.util.List;

@Controller
public class WebController
{
    private UserService userService;
    private RecipeService recipeService;
    private InventoryService inventoryService;
    private ShoppingListService shoppingListService;
    private AuthService authService;

    @Autowired
    public WebController(
            UserService userService,
            RecipeService recipeService,
            InventoryService inventoryService,
            ShoppingListService shoppingListService,
            AuthService authService)
    {
        this.userService = userService;
        this.recipeService = recipeService;
        this.inventoryService = inventoryService;
        this.shoppingListService = shoppingListService;
        this.authService = authService;
    }

    @GetMapping("/login")
    public String getLogin()
    {
        return "login.html";
    }

    @GetMapping("/")
    public String home(Model model)
    {
        model.addAttribute("user", authService.getLoggedInUser().getUsername());
        model.addAttribute("recentlyViewed", recipeService.getRecentlyViewed());

        return "home.html";
    }

    @GetMapping("/signup")
    public String signup()
    {
        return "signup.html";
    }

    @GetMapping("/recipe")
    public String recipe(@RequestParam Long id, Model model)
    {
        try
        {
            Recipe recipe = recipeService.getRecipe(id);
            User user = authService.getLoggedInUser();

            recipeService.addRecentlyViewed(recipe);

            for (Ingredient ingredient : recipe.getIngredients())
            {
                ingredient.setInShoppingList(
                        shoppingListService.isInShoppingList(ingredient)
                );

                ingredient.setInventoryItems(
                        inventoryService.getSimilarIngredients(ingredient)
                );
            }

            model.addAttribute("recipe", recipe);
            model.addAttribute("recentlyViewed", recipeService.getRecentlyViewed());
            model.addAttribute("user", user.getUsername());
            model.addAttribute("ingredients", inventoryService.getIngredients());
            model.addAttribute("displayInstructions", user.isNewUser());

            return "recipe.html";
        }
        catch (RecipeDoesNotExistException e)
        {
            return "invalid_recipe.html";
        }
    }

    @GetMapping("/reset-password")
    public String resetPassword()
    {
        return "reset_password.html";
    }

    @GetMapping("/change-password")
    public String changePassword(@RequestParam String token)
    {
        try
        {
            userService.processPasswordResetToken(token);

            return "change_password.html";
        }
        catch (InvalidPasswordTokenException e)
        {
            return "invalid_token.html";
        }
    }

    // TODO this could have a better name
    @GetMapping("/change-password-logged-in")
    public String changePasswordLoggedIn(Model model)
    {
        model.addAttribute("user", authService.getLoggedInUser().getUsername());
        model.addAttribute("recentlyViewed", recipeService.getRecentlyViewed());

        return "change_password_logged_in.html";
    }

    @GetMapping("/delete-account")
    public String deleteAccount(Model model)
    {
        model.addAttribute("user", authService.getLoggedInUser().getUsername());
        model.addAttribute("recentlyViewed", recipeService.getRecentlyViewed());

        return "delete_account.html";
    }

    @GetMapping("/shopping-list")
    public String shoppingList(Model model)
    {
        model.addAttribute("user", authService.getLoggedInUser().getUsername());
        model.addAttribute("recentlyViewed", recipeService.getRecentlyViewed());
        model.addAttribute("shoppingList", shoppingListService.getShoppingList());
        model.addAttribute("ingredients", inventoryService.getIngredients());

        return "shopping_list.html";
    }

    @GetMapping("/inventory")
    public String inventory(Model model)
    {
        model.addAttribute("user", authService.getLoggedInUser().getUsername());
        model.addAttribute("recentlyViewed", recipeService.getRecentlyViewed());
        model.addAttribute("inventory", inventoryService.getInventory());
        model.addAttribute("ingredients", inventoryService.getIngredients());

        return "inventory.html";
    }
}

File: src/main/kotlin/thecookingpot/recipe/controller/WebController.kt
package thecookingpot.recipe.controller

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestParam
import thecookingpot.recipe.exception.InvalidPasswordTokenException
import thecookingpot.recipe.exception.RecipeDoesNotExistException
import thecookingpot.recipe.service.*

@Controller
class WebController // TODO I should have a custom error page
@Autowired constructor(
        private val userService: UserService,
        private val recipeService: RecipeService,
        private val inventoryService: InventoryService,
        private val shoppingListService: ShoppingListService,
        private val authService: AuthService) {
    @GetMapping("/login")
    fun login(): String {
        return "login.html"
    }

    @GetMapping("/")
    fun home(model: Model): String {
        model.apply {
            addAttribute("user", authService.loggedInUser.username)
            addAttribute("recentlyViewed", recipeService.recentlyViewed)
        }
        
        return "home.html"
    }

    // TODO unfinished react implementation
    @GetMapping("/react")
    fun homeReact(model: Model): String {
        model.apply {
            addAttribute("loggedInUser", authService.loggedInUser.username)
            addAttribute("recipes", recipeService.recipes)
            addAttribute("recentlyViewedRecipes", recipeService.recentlyViewed)
        }

        return "home_react.html"
    }

    @GetMapping("/signup")
    fun signup(): String {
        return "signup.html"
    }

    @GetMapping("/recipe")
    fun recipe(@RequestParam id: Long, model: Model): String {
        return try {
            val recipe = recipeService.getRecipe(id)

            recipeService.addRecentlyViewed(recipe)
            
            recipe.ingredients.forEach { ingredient ->
                ingredient.isInShoppingList = shoppingListService.isInShoppingList(ingredient)
                ingredient.inventoryItems = inventoryService.getSimilarIngredients(ingredient)
            }
            
            model.apply {
                addAttribute("recipe", recipe)
                addAttribute("recentlyViewed", recipeService.recentlyViewed)
                addAttribute("user", authService.loggedInUser.username)
                addAttribute("ingredients", inventoryService.ingredients)
                addAttribute("displayInstructions", authService.loggedInUser.newUser)
            }
            
            "recipe.html"
        } catch (e: RecipeDoesNotExistException) {
            "invalid-recipe.html"
        }
    }

    @GetMapping("/reset-password")
    fun resetPassword(): String {
        return "reset-password.html"
    }

    @GetMapping("/change-password-with-token")
    fun changePassword(@RequestParam token: String): String {
        return try {
            userService.processPasswordResetToken(token)
            "change-password-with-token.html"
        } catch (e: InvalidPasswordTokenException) {
            "invalid-token.html"
        }
    }

    @GetMapping("/change-password")
    fun changePasswordLoggedIn(model: Model): String {
        model.apply {
            addAttribute("loggedInUser", authService.loggedInUser.username)
            addAttribute("recentlyViewedRecipes", recipeService.recentlyViewed)
        }

        return "change-password.html"
    }

    @GetMapping("/delete-account")
    fun deleteAccount(model: Model): String {
        model.apply {
            addAttribute("user", authService.loggedInUser.username)
            addAttribute("recentlyViewed", recipeService.recentlyViewed)
        }

        return "delete-account.html"
    }

    @GetMapping("/shopping-list")
    fun shoppingList(model: Model): String {
        model.apply {
            addAttribute("user", authService.loggedInUser.username)
            addAttribute("recentlyViewed", recipeService.recentlyViewed)
            addAttribute("shoppingList", shoppingListService.shoppingList)
            addAttribute("ingredients", inventoryService.ingredients)
        }

        return "shopping-list.html"
    }

    @GetMapping("/inventory")
    fun inventory(model: Model): String {
        model.apply {
            addAttribute("user", authService.loggedInUser.username)
            addAttribute("recentlyViewed", recipeService.recentlyViewed)
            addAttribute("inventory", inventoryService.inventory)
            addAttribute("ingredients", inventoryService.ingredients)
        }

        return "inventory.html"
    }
}